File: backend/internal/api/stations.go
package api

import (
	"encoding/json"
	"log/slog"
	"net/http"
	"net/url"
	"strconv"
	"strings"

	"github.com/go-chi/chi/v5"

	"opencharge/internal/domain"
	"opencharge/internal/repository"
)

type StationsHandler struct {
	Stations *repository.StationRepository
	Tariffs  *repository.TariffRepository
}

func NewStationsHandler(stations *repository.StationRepository, tariffs *repository.TariffRepository) *StationsHandler {
	return &StationsHandler{Stations: stations, Tariffs: tariffs}
}

// ListStations handles GET /stations?bbox=minLng,minLat,maxLng,maxLat&operator=&hasTariffs=&source=&connectorType=&minPowerKw=&minPriceCentsPerKwh=&maxPriceCentsPerKwh=&chargeKWh=&chargeMinutes=&excludeSubscriptionPlans=&limit=&offset=
// It never loads the whole dataset: bbox is mandatory, and the map/frontend
// is expected to re-query on every viewport change.
//
// source accepts a comma-separated list of "source:plan" pairs (e.g.
// "izivia:standard,electra:subscription"); a bare source name (no ":plan")
// defaults to the "standard" plan. It never filters stations out of the
// response: it only controls which (source, plan) pairs
// selectedSourcesPricing is computed from, so the map can gray out
// stations lacking a tariff from the selection instead of hiding them.
//
// excludeSubscriptionPlans=true drops any tariff on the
// domain.TariffPlanSubscription plan from both pricingSummary and
// selectedSourcesPricing, so the price shown for a station never assumes a
// paid subscription the caller may not have.
//
// chargeKWh/chargeMinutes, when given alongside min/maxPriceCentsPerKwh,
// switch the price-range filter to the estimated TOTAL cost (in cents) of a
// session delivering chargeKWh over chargeMinutes (energy + any per-minute
// rate + any flat session fee) instead of a plain €/kWh rate — see
// domain.StationFilter.MinPriceCentsPerKWh's doc comment.
func (h *StationsHandler) ListStations(w http.ResponseWriter, r *http.Request) {
	q := r.URL.Query()

	bboxParam := q.Get("bbox")
	if bboxParam == "" {
		writeError(w, http.StatusBadRequest, "bbox query parameter is required (minLng,minLat,maxLng,maxLat)")
		return
	}
	parts := strings.Split(bboxParam, ",")
	if len(parts) != 4 {
		writeError(w, http.StatusBadRequest, "bbox must have 4 comma-separated values: minLng,minLat,maxLng,maxLat")
		return
	}
	coords := make([]float64, 4)
	for i, p := range parts {
		v, err := strconv.ParseFloat(strings.TrimSpace(p), 64)
		if err != nil {
			writeError(w, http.StatusBadRequest, "bbox contains an invalid number: "+p)
			return
		}
		coords[i] = v
	}

	filter := domain.StationFilter{
		MinLng:   coords[0],
		MinLat:   coords[1],
		MaxLng:   coords[2],
		MaxLat:   coords[3],
		Operator: q.Get("operator"),
		Sources:  parseSourcePlanPairs(q.Get("source")),
	}
	if v := q.Get("hasTariffs"); v != "" {
		b, err := strconv.ParseBool(v)
		if err == nil {
			filter.HasTariffs = &b
		}
	}
	if v := q.Get("connectorType"); v != "" {
		filter.ConnectorTypes = strings.Split(v, ",")
	}
	if v := q.Get("minPowerKw"); v != "" {
		if p, err := strconv.ParseFloat(v, 64); err == nil {
			filter.MinPowerKW = &p
		}
	}
	if v := q.Get("minPriceCentsPerKwh"); v != "" {
		if p, err := strconv.ParseFloat(v, 64); err == nil {
			filter.MinPriceCentsPerKWh = &p
		}
	}
	if v := q.Get("maxPriceCentsPerKwh"); v != "" {
		if p, err := strconv.ParseFloat(v, 64); err == nil {
			filter.MaxPriceCentsPerKWh = &p
		}
	}
	if v := q.Get("chargeKWh"); v != "" {
		if p, err := strconv.ParseFloat(v, 64); err == nil {
			filter.ChargeKWh = &p
		}
	}
	if v := q.Get("chargeMinutes"); v != "" {
		if p, err := strconv.ParseFloat(v, 64); err == nil {
			filter.ChargeMinutes = &p
		}
	}
	if v := q.Get("excludeSubscriptionPlans"); v != "" {
		if b, err := strconv.ParseBool(v); err == nil {
			filter.ExcludeSubscriptionPlans = b
		}
	}
	if v := q.Get("limit"); v != "" {
		if n, err := strconv.Atoi(v); err == nil {
			filter.Limit = n
		}
	}
	if v := q.Get("offset"); v != "" {
		if n, err := strconv.Atoi(v); err == nil {
			filter.Offset = n
		}
	}

	summaries, err := h.Stations.ListByBBox(r.Context(), filter)
	if err != nil {
		slog.Error("list stations", "error", err)
		writeError(w, http.StatusInternalServerError, "failed to list stations")
		return
	}

	items := make([]stationListItemDTO, 0, len(summaries))
	for _, s := range summaries {
		items = append(items, toStationListItemDTO(s))
	}
	writeJSON(w, http.StatusOK, items)
}

// ListSources handles GET /sources: every tariff source currently ingested,
// each with its available price plans (e.g. "izivia" -> ["standard"],
// "electra" -> ["app", "public", "subscription"]), so the frontend can
// build its operator filter and plan selector dynamically instead of
// hardcoding anything.
func (h *StationsHandler) ListSources(w http.ResponseWriter, r *http.Request) {
	sources, err := h.Tariffs.ListDistinctSourcesWithPlans(r.Context())
	if err != nil {
		slog.Error("list sources", "error", err)
		writeError(w, http.StatusInternalServerError, "failed to list sources")
		return
	}
	items := make([]sourcePlansDTO, 0, len(sources))
	for _, s := range sources {
		items = append(items, sourcePlansDTO{ID: s.Source, Plans: s.Plans})
	}
	writeJSON(w, http.StatusOK, items)
}

// GetStation handles GET /stations/{id} where id is "irve:<id_pdc_itinerance>".
func (h *StationsHandler) GetStation(w http.ResponseWriter, r *http.Request) {
	rawID := chi.URLParam(r, "id")
	// chi does not percent-decode route params, and clients correctly
	// encode the ":" in "irve:<id>" (e.g. via encodeURIComponent), so
	// decode before matching the "irve:" prefix.
	if decoded, err := url.PathUnescape(rawID); err == nil {
		rawID = decoded
	}
	irveID := strings.TrimPrefix(rawID, "irve:")
	if irveID == "" {
		writeError(w, http.StatusBadRequest, "missing station id")
		return
	}

	station, err := h.Stations.GetByIRVEID(r.Context(), irveID)
	if err != nil {
		slog.Error("get station", "error", err)
		writeError(w, http.StatusInternalServerError, "failed to get station")
		return
	}
	if station == nil {
		writeError(w, http.StatusNotFound, "station not found")
		return
	}

	tariffs, err := h.Tariffs.ListByStation(r.Context(), station.ID)
	if err != nil {
		slog.Error("list tariffs", "error", err)
		writeError(w, http.StatusInternalServerError, "failed to load tariffs")
		return
	}

	tariffDTOs := make([]tariffDTO, 0, len(tariffs))
	for _, t := range tariffs {
		tariffDTOs = append(tariffDTOs, toTariffDTO(t))
	}

	writeJSON(w, http.StatusOK, stationDetailResponse{
		Station: toStationDetailDTO(*station),
		Tariffs: tariffDTOs,
	})
}

func writeJSON(w http.ResponseWriter, status int, payload any) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	_ = json.NewEncoder(w).Encode(payload)
}

func writeError(w http.ResponseWriter, status int, message string) {
	writeJSON(w, status, map[string]string{"error": message})
}

// parseSourcePlanPairs splits a comma-separated "source" query param (e.g.
// "izivia,electra:subscription") into normalized "source:plan" pairs,
// defaulting a bare source name to the standard plan. Dropping empty
// entries and whitespace. An empty input returns nil.
func parseSourcePlanPairs(raw string) []string {
	if raw == "" {
		return nil
	}
	parts := strings.Split(raw, ",")
	pairs := make([]string, 0, len(parts))
	for _, p := range parts {
		p = strings.TrimSpace(p)
		if p == "" {
			continue
		}
		if !strings.Contains(p, ":") {
			p = p + ":" + domain.TariffPlanStandard
		}
		pairs = append(pairs, p)
	}
	return pairs
}

File: frontend/web/src/api/stations.js
const API_BASE = import.meta.env.VITE_API_BASE_URL || "http://localhost:8080";

/**
 * Throws a structured error for a failed response: callers can branch on
 * `err.status` (e.g. friendlyFetchErrorMessage in utils/format.js) instead
 * of parsing the message string, and a missing status (network failure,
 * CORS, DNS) stays distinguishable from a real HTTP error status.
 */
async function throwForStatus(res, what) {
  const err = new Error(`${what} failed: ${res.status}`);
  err.status = res.status;
  throw err;
}

/**
 * Fetch stations intersecting a map viewport. bbox is mandatory: the map
 * never loads the whole IRVE dataset, only what's currently visible.
 *
 * `sources` (array of "source:plan" pairs, e.g. ["izivia:standard",
 * "electra:subscription"] — see utils/pricing.js#sourcePlanPairs) never
 * filters stations out server-side: it only asks the API to compute
 * `selectedSourcesPricing` for those (source, plan) pairs, so stations
 * without a tariff from the selection can still be shown, just without a
 * price.
 *
 * excludeSubscriptionPlans, when true, drops subscription-plan tariffs from
 * both pricingSummary and selectedSourcesPricing server-side (see backend
 * GET /stations docs), so the price used for the map marker never assumes
 * a paid subscription.
 *
 * chargeKWh/chargeMinutes, when given alongside min/maxPriceCentsPerKwh,
 * switch the price-range filter server-side to the total cost of a session
 * of that size (energy + time + flat fee) instead of a plain €/kWh rate —
 * see backend GET /stations docs.
 */
export async function fetchStationsInBBox(
  bbox,
  {
    operator,
    hasTariffs,
    sources,
    connectorTypes,
    minPowerKw,
    minPriceCentsPerKwh,
    maxPriceCentsPerKwh,
    chargeKWh,
    chargeMinutes,
    excludeSubscriptionPlans,
    limit,
    signal,
  } = {}
) {
  const params = new URLSearchParams();
  params.set("bbox", bbox.join(","));
  if (operator) params.set("operator", operator);
  if (hasTariffs !== undefined) params.set("hasTariffs", String(hasTariffs));
  if (sources && sources.length > 0) params.set("source", sources.join(","));
  if (connectorTypes && connectorTypes.length > 0) params.set("connectorType", connectorTypes.join(","));
  if (minPowerKw != null) params.set("minPowerKw", String(minPowerKw));
  if (minPriceCentsPerKwh != null) params.set("minPriceCentsPerKwh", String(minPriceCentsPerKwh));
  if (maxPriceCentsPerKwh != null) params.set("maxPriceCentsPerKwh", String(maxPriceCentsPerKwh));
  if (chargeKWh != null) params.set("chargeKWh", String(chargeKWh));
  if (chargeMinutes != null) params.set("chargeMinutes", String(chargeMinutes));
  if (excludeSubscriptionPlans) params.set("excludeSubscriptionPlans", "true");
  params.set("limit", String(limit ?? 500));

  const res = await fetch(`${API_BASE}/stations?${params.toString()}`, { signal });
  if (!res.ok) await throwForStatus(res, "GET /stations");
  return res.json();
}

export async function fetchStationDetails(id, { signal } = {}) {
  const res = await fetch(`${API_BASE}/stations/${encodeURIComponent(id)}`, { signal });
  if (!res.ok) await throwForStatus(res, `GET /stations/${id}`);
  return res.json();
}

/**
 * Fetch every tariff source currently known to the backend (e.g.
 * ["electra", "izivia"]), so the operator filter never needs a hardcoded
 * list and picks up new sources as soon as they're ingested.
 */
export async function fetchSources({ signal } = {}) {
  const res = await fetch(`${API_BASE}/sources`, { signal });
  if (!res.ok) await throwForStatus(res, "GET /sources");
  return res.json();
}

/**
 * A site's live is_available status, via our own backend (see
 * components/FreshmileAvailability.jsx) rather than calling Freshmile's API
 * directly from the browser: that direct call is blocked by CORS in
 * production (Freshmile's API sends no Access-Control-Allow-Origin header —
 * confirmed against the real deployment), so the backend proxies it
 * (GET /freshmile/availability/{locationId} — see backend api/freshmile.go).
 */
export async function fetchFreshmileAvailability(locationId, { signal } = {}) {
  const res = await fetch(`${API_BASE}/freshmile/availability/${encodeURIComponent(locationId)}`, { signal });
  if (!res.ok) await throwForStatus(res, `GET /freshmile/availability/${locationId}`);
  return res.json();
}

/**
 * A nationwide-average fuel price (SP95-E10 today), for the
 * essence/électrique cost comparison (see utils/pricing.js#
 * thermalEquivalentCost) — already averaged and cached server-side (see
 * backend api/fuelprice.go), so this is cheap to call once per session.
 */
export async function fetchFuelPrice({ signal } = {}) {
  const res = await fetch(`${API_BASE}/fuel-price`, { signal });
  if (!res.ok) await throwForStatus(res, "GET /fuel-price");
  return res.json();
}
